### PR TITLE
Auto connect sessions on program startup

### PR DESCRIPTION
**Pull request by wyfinger**

### CODE REVIEW

**Comment by kostapc** [comment]
@wyfinger thank you again! :)
I'll check it later, because have some ideas about autoconnect and reconnect on fail.

**Comment by wyfinger** [comment]
Ok. I did it as a concept and for my own needs.

File: TunnelManager/Core.cs [hunk 1]
```diff
@@ -57,7 +58,29 @@ private void Initialize()
...
+                if (session.AutoStart && !session.IsOpen)
+                {
+                    // auto-start session on program startup
+                    try
+                    {
+                        session.Open();
+                    }
+                    catch (SessionAlreadyOpenException)
+                    {
+                        MessageBox.Show("Session already open.");
```
Comment (kostapc) [inline TunnelManager/Core.cs]: MessageBox maybe annoying if something went wrong.
I want to propose use **UserNotifications.Notify(this.session.Name, "<message>");** that i want to design special for any user notifications.

File: TunnelManager/Forms/SettingsForm.cs [hunk 2]
```diff
@@ -41,8 +41,8 @@ public SettingsForm()
 
         private void SettingsForm_Shown(object sender, EventArgs e)
         {
-            Core.Instance().Refresh();
-
+            //Core.Instance().Refresh();  // problem if clear opened sessions on settings form show
```
Comment (kostapc) [inline TunnelManager/Forms/SettingsForm.cs]: What kind of problem?

File: TunnelManager/Session.cs [hunk 3]
```diff
@@ -229,6 +237,7 @@ public virtual void Serialize()
                 RegistryKey ptmSessionKey = Registry.CurrentUser.CreateSubKey(this.PuttyTunnelManagerKeyPath);
                 ptmSessionKey.SetValue(PTM_REGISTRY_KEY_SESSION_PORTFORWARDINGS, buffer.ToString(), RegistryValueKind.String);
                 ptmSessionKey.SetValue(PTM_REGISTRY_KEY_SESSION_USEPTMFORTUNNELS, this.usePtmForTunnels, RegistryValueKind.DWord);
+                ptmSessionKey.SetValue(PTM_REGISTRY_KEY_SESSION_AUTOSTART, this.autoStart, RegistryValueKind.DWord);
```
Comment (kostapc) [inline TunnelManager/Session.cs]: If sessions not stored in PMT - autostart not available? If so "autostart" option must be disabled. In other case autostart function must work Independent.

**Comment by kostapc** [comment]
ok, i see you don't going to do anything with it :(
I check code and accept it as is for future tiny refactoring.

**Comment by wyfinger** [comment]
Sorry, there is absolutely no time and I'm also quite new to C #.
What about Core.Instance (). Refresh (); The matter is that it recreates connections.
In general, we need to seriously review all the code, because if we kill ptman.exe, plink.exe processes will remain running and later ptman will not be able to create a connection. Probably we need to go through the list of running processes, look for plink and look at their parameters.

**Comment by kostapc** [comment]
I agree, there many problems with program restarting, wake up after hibernation and autostart.

**Pull request merged**